ntoskrnl/io/debug.c

```diff
@@ -0,0 +1,561 @@
+        }
+
+        AltList = (PIO_RESOURCE_LIST)(AltList->Descriptors + AltList->Count);
+        DPRINT1("End Descriptors %p\n", AltList);
+    }
+}
+
+/* Gets the name of the device node state. */
+
+PWSTR
+NTAPI
+PipGetDeviceNodeStateName(
+    _In_ PNP_DEVNODE_STATE State)
+{
+    PWSTR StateName;
+
+    switch (State)
+    {
+        case DeviceNodeUnspecified:
+            StateName = L"DeviceNodeUnspecified";
+            break;
+
+        case DeviceNodeUninitialized:
+            StateName = L"DeviceNodeUninitialized";
+            break;
+
+        case DeviceNodeInitialized:
+            StateName = L"DeviceNodeInitialized";
+            break;
+
+        case DeviceNodeDriversAdded:
+            StateName = L"DeviceNodeDriversAdded";
+            break;
+
+        case DeviceNodeResourcesAssigned:
+            StateName = L"DeviceNodeResourcesAssigned";
+            break;
+
+        case DeviceNodeStartPending:
+            StateName = L"DeviceNodeStartPending";
+            break;
+
+        case DeviceNodeStartCompletion:
+            StateName = L"DeviceNodeStartCompletion";
+            break;
+
+        case DeviceNodeStartPostWork:
+            StateName = L"DeviceNodeStartPostWork";
+            break;
+
+        case DeviceNodeStarted:
+            StateName = L"DeviceNodeStarted";
+            break;
+
+        case DeviceNodeQueryStopped:
+            StateName = L"DeviceNodeQueryStopped";
+            break;
+
+        case DeviceNodeStopped:
+            StateName = L"DeviceNodeStopped";
+            break;
+
+        case DeviceNodeRestartCompletion:
+            StateName = L"DeviceNodeRestartCompletion";
+            break;
+
+        case DeviceNodeEnumeratePending:
+            StateName = L"DeviceNodeEnumeratePending";
+            break;
+
+        case DeviceNodeEnumerateCompletion:
+            StateName = L"DeviceNodeEnumerateCompletion";
+            break;
+
+        case DeviceNodeAwaitingQueuedDeletion:
+            StateName = L"DeviceNodeAwaitingQueuedDeletion";
+            break;
+
+        case DeviceNodeAwaitingQueuedRemoval:
+            StateName = L"DeviceNodeAwaitingQueuedRemoval";
+            break;
+
+        case DeviceNodeQueryRemoved:
+            StateName = L"DeviceNodeQueryRemoved";
+            break;
+
+        case DeviceNodeRemovePendingCloses:
+            StateName = L"DeviceNodeRemovePendingCloses";
+            break;
+
+        case DeviceNodeRemoved:
+            StateName = L"DeviceNodeRemoved";
+            break;
+
+        case DeviceNodeDeletePendingCloses:
+            StateName = L"DeviceNodeDeletePendingCloses";
+            break;
+
+        case DeviceNodeDeleted:
+            StateName = L"DeviceNodeDeleted";
+            break;
+
+        default:
+            if (State != MaxDeviceNodeState)
+            {
+                DPRINT1("PipGetDeviceNodeStateName: State %X\n", State);
+            }
+            break;
```

Comment:
Note that in case the "default" is taken, the returned StateName value is undefined.

Reply:
Changed.